// ts-simple-ast/dist/compiler/base/AsyncableNode.js
"use strict";
var __extends = (this && this.__extends)/* istanbul ignore next */ || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var errors = require("./../../errors");
var callBaseFill_1 = require("./../callBaseFill");
function AsyncableNode(Base) {
    return /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.isAsync = function () {
            return this.hasModifier(ts.SyntaxKind.AsyncKeyword);
        };
        class_1.prototype.getAsyncKeyword = function () {
            return this.getFirstModifierByKind(ts.SyntaxKind.AsyncKeyword);
        };
        class_1.prototype.getAsyncKeywordOrThrow = function () {
            return errors.throwIfNullOrUndefined(this.getFirstModifierByKind(ts.SyntaxKind.AsyncKeyword), "Expected to find an async keyword.");
        };
        class_1.prototype.setIsAsync = function (value) {
            this.toggleModifier("async", value);
            return this;
        };
        class_1.prototype.fill = function (structure) {
            callBaseFill_1.callBaseFill(Base.prototype, this, structure);
            if (structure.isAsync != null)
                this.setIsAsync(structure.isAsync);
            return this;
        };
        return class_1;
    }(Base));
}
exports.AsyncableNode = AsyncableNode;
