// ts-simple-ast/dist/compiler/base/QuestionTokenableNode.js
"use strict";
var __extends = (this && this.__extends)/* istanbul ignore next */ || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var errors = require("./../../errors");
var manipulation_1 = require("./../../manipulation");
var callBaseFill_1 = require("./../callBaseFill");
function QuestionTokenableNode(Base) {
    return /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.hasQuestionToken = function () {
            return this.compilerNode.questionToken != null;
        };
        class_1.prototype.getQuestionTokenNode = function () {
            return this.getNodeFromCompilerNodeIfExists(this.compilerNode.questionToken);
        };
        class_1.prototype.getQuestionTokenNodeOrThrow = function () {
            return errors.throwIfNullOrUndefined(this.getQuestionTokenNode(), "Expected to find a question token.");
        };
        class_1.prototype.setHasQuestionToken = function (value) {
            var questionTokenNode = this.getQuestionTokenNode();
            var hasQuestionToken = questionTokenNode != null;
            if (value === hasQuestionToken)
                return this;
            if (value) {
                var colonNode = this.getFirstChildByKindOrThrow(ts.SyntaxKind.ColonToken);
                manipulation_1.insertIntoParent({
                    insertPos: colonNode.getStart(),
                    childIndex: colonNode.getChildIndex(),
                    insertItemsCount: 1,
                    parent: this,
                    newText: "?"
                });
            }
            else {
                manipulation_1.removeChildren({ children: [questionTokenNode] });
            }
            return this;
        };
        class_1.prototype.fill = function (structure) {
            callBaseFill_1.callBaseFill(Base.prototype, this, structure);
            if (structure.hasQuestionToken != null)
                this.setHasQuestionToken(structure.hasQuestionToken);
            return this;
        };
        return class_1;
    }(Base));
}
exports.QuestionTokenableNode = QuestionTokenableNode;
