// ts-simple-ast/dist/utils/printNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var newLineKindToTs_1 = require("./newLineKindToTs");
var ManipulationSettings_1 = require("./../ManipulationSettings");
function printNode(node, sourceFileOrOptions, secondOverloadOptions) {
    var isFirstOverload = sourceFileOrOptions == null || sourceFileOrOptions.kind !== ts.SyntaxKind.SourceFile;
    var options = getOptions();
    var sourceFile = getSourceFile();
    var printer = ts.createPrinter({
        newLine: newLineKindToTs_1.newLineKindToTs(options.newLineKind || ManipulationSettings_1.NewLineKind.LineFeed),
        removeComments: options.removeComments || false
    });
    if (sourceFile == null)
        return printer.printFile(node);
    else
        return printer.printNode(options.emitHint || ts.EmitHint.Unspecified, node, sourceFile);
    function getSourceFile() {
        if (isFirstOverload) {
            if (node.kind === ts.SyntaxKind.SourceFile)
                return undefined;
            var scriptKind = getScriptKind();
            return ts.createSourceFile("print." + getFileExt(scriptKind), "", ts.ScriptTarget.Latest, false, scriptKind);
        }
        return sourceFileOrOptions;
        function getScriptKind() {
            return options.scriptKind == null ? ts.ScriptKind.TSX : options.scriptKind;
        }
        function getFileExt(scriptKind) {
            if (scriptKind === ts.ScriptKind.JSX || scriptKind === ts.ScriptKind.TSX)
                return "tsx";
            return "ts";
        }
    }
    function getOptions() {
        return (isFirstOverload ? sourceFileOrOptions : secondOverloadOptions) || {};
    }
}
exports.printNode = printNode;
