// ts-simple-ast/dist/main.js
"use strict";
function __export(m) {
    for (var p in m) /* istanbul ignore else */ if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./compiler"));
var TsSimpleAst_1 = require("./TsSimpleAst");
exports.default = TsSimpleAst_1.TsSimpleAst;
var fileSystem_1 = require("./fileSystem");
exports.Directory = fileSystem_1.Directory;
exports.DirectoryEmitResult = fileSystem_1.DirectoryEmitResult;
__export(require("./ManipulationSettings"));
var createWrappedNode_1 = require("./createWrappedNode");
exports.createWrappedNode = createWrappedNode_1.createWrappedNode;
var getCompilerOptionsFromTsConfig_1 = require("./utils/tsconfig/getCompilerOptionsFromTsConfig");
exports.getCompilerOptionsFromTsConfig = getCompilerOptionsFromTsConfig_1.getCompilerOptionsFromTsConfig;
var printNode_1 = require("./utils/printNode");
exports.printNode = printNode_1.printNode;
var TypeGuards_1 = require("./utils/TypeGuards");
exports.TypeGuards = TypeGuards_1.TypeGuards;
